### ISSUE TITLE
env_posix.c:125: error: storage size of ‘ai_hints’ isn’t known

### ISSUE THREAD

**Issue by vnart** [mdirolf/nginx-gridfs#47]
-o objs/addon/src/env_posix.o \
        /root/nginx_mongo/mongo-c-driver/src/env_posix.c
/root/nginx_mongo/mongo-c-driver/src/env_posix.c: In function ‘mongo_socket_connect’:
/root/nginx_mongo/mongo-c-driver/src/env_posix.c:125: error: storage size of ‘ai_hints’ isn’t known
/root/nginx_mongo/mongo-c-driver/src/env_posix.c:142: warning: implicit declaration of function ‘getaddrinfo’
/root/nginx_mongo/mongo-c-driver/src/env_posix.c:144: warning: implicit declaration of function ‘gai_strerror’
/root/nginx_mongo/mongo-c-driver/src/env_posix.c:149: error: dereferencing pointer to incomplete type
/root/nginx_mongo/mongo-c-driver/src/env_posix.c:150: error: dereferencing pointer to incomplete type
/root/nginx_mongo/mongo-c-driver/src/env_posix.c:150: error: dereferencing pointer to incomplete type
/root/nginx_mongo/mongo-c-driver/src/env_posix.c:150: error: dereferencing pointer to incomplete type
/root/nginx_mongo/mongo-c-driver/src/env_posix.c:156: error: dereferencing pointer to incomplete type
/root/nginx_mongo/mongo-c-driver/src/env_posix.c:156: error: dereferencing pointer to incomplete type
/root/nginx_mongo/mongo-c-driver/src/env_posix.c:163: error: dereferencing pointer to incomplete type
/root/nginx_mongo/mongo-c-driver/src/env_posix.c:176: warning: implicit declaration of function ‘freeaddrinfo’
/root/nginx_mongo/mongo-c-driver/src/env_posix.c:125: warning: unused variable ‘ai_hints’
make[1]: **\* [objs/addon/src/env_posix.o] Error 1
make[1]: Leaving directory `/root/nginx-1.2.2'
make: **\* [build] Error 2

**Comment by latticekun** [comment]
I also encountered this problem, can you get the answer？

**Comment by vnart** [comment]
Yes, I build the mongo-c-driver first then build nginx with the mongo-c-drive lib

**Comment by latticekun** [comment]
thanks，I will try it...

**Comment by johnbinwang** [comment]
Could you please explain your method in detail? I am also encountered this
problem.

On Mon, Aug 27, 2012 at 11:20 AM, vnart notifications@github.com wrote:

> Yes, I build the mongo-c-driver first then build nginx with the
> mongo-c-drive lib
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/mdirolf/nginx-gridfs/issues/47#issuecomment-8046312.

## 

cheers,
Johnbin Wang

**Comment by vnart** [comment]
use make or scons to build the mongo-c-drive lib ( refer to their website for further instruction). Then copy the lib to the right location.

Modify config for nginx-gridfs to:
NGX_ADDON_SRCS="$NGX_ADDON_SRCS $ngx_addon_dir/ngx_http_gridfs_module.c  "
NGX_ADDON_DEPS="$NGX_ADDON_DEPS $ngx_addon_dir/mongo-c-driver/src/*.h"

CFLAGS="$CFLAGS --std=c99 -I src -I /usr/local/include -L /usr/local/lib -lmongoc "

Then build nginx with --add-module=/root/july2012/gridfs/ --with-ld-opt=-lmongoc

Should give you no error at all. However, the current version of nginx-grifs does not work well with replica set when the primary went offline. I had to make some change to the nginx-grifs source code to make it force reconnect the connection to detect the primary has change.

One more thing, have to use nginx restart for nginx-grifs to work other. If use reload it will cause problem and nginx just quit.

**Comment by latticekun** [comment]
it was done by your way! it works well! thank you very much! @vnart

**Comment by johnbinwang** [comment]
Thank you vnart, it works! The images fetched by nginx-grifs can be cached
now.

On Mon, Aug 27, 2012 at 4:07 PM, vnart notifications@github.com wrote:

> use make or scons to build the mongo-c-drive lib ( refer to their website
> for further instruction). Then copy the lib to the right location.
> 
> Modify config for nginx-gridfs to:
> NGX_ADDON_SRCS="$NGX_ADDON_SRCS $ngx_addon_dir/ngx_http_gridfs_module.c "
> NGX_ADDON_DEPS="$NGX_ADDON_DEPS $ngx_addon_dir/mongo-c-driver/src/*.h"
> 
> CFLAGS="$CFLAGS --std=c99 -I src -I /usr/local/include -L /usr/local/lib
> -lmongoc "
> 
> Then build nginx with --add-module=/root/july2012/gridfs/
> --with-ld-opt=-lmongoc
> 
> Should give you no error at all. However, the current version of
> nginx-grifs does not work well with replica set when the primary went
> offline. I had to make some change to the nginx-grifs source code to make
> it force reconnect the connection to detect the primary has change.
> 
> One more thing, have to use nginx restart for nginx-grifs to work other.
> If use reload it will cause problem and nginx just quit.
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/mdirolf/nginx-gridfs/issues/47#issuecomment-8049427.

## 

cheers,
Johnbin Wang
